### PR TITLE
LC::Check: fix link() for hardlink creation

### LINKED ISSUE #17: LC::Check::link 17.2: hardlink management is buggy
The current implementation of `LC::Check` is buggy with respect to `ln` when creating a hardlink on a symlink.
* `ln` allows to create a  hardlink on a symlink whatever its target is, including a broken symlink or a direcotry, as demonstrated by the following example:
```
lx3/jouvin % ln -s missing_dir /tmp/broken_symlink
lx3/jouvin % ln /tmp/broken_symlink /tmp/broken_hardlink                                                                                                                  
lx3/jouvin % ls -i /tmp/broken_*
216 /tmp/broken_hardlink  216 /tmp/broken_symlink
```
* `LC::Check::symlink` checks that the target exists and is not a directory, which is too restrictive and inconsistent with `ls` (a comment suggests that it was intentional at a time where `ln` was not allowing it). There is no possible workaround for this problem: if not fixed, `CAF::Path` (https://github.com/quattor/CAF/pull/225) should forbid it explicitely.
* `LC::Check::symlink`  incorrectly checks the target existence if the target is a relative path: it does the same test as for a symlink (where a relative path for the symlink target is interpreted relative to the symlink parent directory) where the test should be done on the target as it was specified. It is possible to work around this problem in `CAF::Path` by converting the relative path to an absolute path before calling  `LC::Path::link`.

### PR DESCRIPTION
**Pull request by jouvin**
Fixes #17.

This fix is potentially needed by https://github.com/quattor/CAF/pull/225 (see comment https://github.com/quattor/CAF/pull/225#issuecomment-284883048)

### CODE REVIEW

**Comment by stdweird** [comment]
running in strace: `strace -f -s 256 -o /tmp/out ln /tmp/broken /tmp/hl_br`
you can see what ln does 
```
27847 stat("/tmp/hl_br", 0x7fff3023a5b0) = -1 ENOENT (No such file or directory)
27847 lstat("/tmp/broken", {st_mode=S_IFLNK|0777, st_size=15, ...}) = 0
27847 link("/tmp/broken", "/tmp/hl_br") = 0
```
this is on el5, so the old note must have referred to an even older EL version.

File: src/main/perl/Check.pm [hunk 1]
```diff
@@ -536,28 +536,28 @@ sub link ($$;%) {
...
-    }
+
     if ($opt{hard}) {
-	# note: we follow symlinks for $target to be compatible with ln(1)
-	@stat = LC::Fatal::stat($test);
+        # Check target that is required to exist.
+	@stat = LC::Fatal::lstat($target);
 	if (@stat) {
-	    if (-d _) {
+	    if (-d _ && ! -l _) {
```
Comment (stdweird) [inline src/main/perl/Check.pm]: with lstat and the magic `_`, it's either `-d` or `-l`, so the `&& ! -l` is not required
```
[stdweird@gast039e tmp]$ perl -e '@stat=lstat("broken");print "D", -d _, "F", -f _, "E", -e _, "L", -l _,"\n"'
DFE1L1
[stdweird@gast039e tmp]$ perl -e '@stat=lstat("dir");print "D", -d _, "F", -f _, "E", -e _, "L", -l _,"\n"'
D1FE1L
```
without `_` magic, it's 
```
[stdweird@gast039e tmp]$ perl -e 'print "D", -d "broken", "F", -f "broken", "E", -e "broken", "L", -l "broken","\n"'
DFEL1
```

best to add comment 😄

File: src/main/perl/Check.pm [hunk 2]
```diff
@@ -584,7 +584,7 @@ sub link ($$;%) {
 	    #
 	    # check hard link
 	    #
-	    if (-d _) {
+	    if (-d _ && ! -l _) {
```
Comment (stdweird) [inline src/main/perl/Check.pm]: also not needed

File: src/main/perl/Check.pm (hunk 1 above)
Comment (jouvin) [inline src/main/perl/Check.pm]: I was about to look at what was the `_` magic doing! In this case this is even better.

Comment (stdweird) [inline src/main/perl/Check.pm]: it's a magic filehandle which has lastest stat or lstat data, so you can run as many tests on it without having to thrash the filesystem.

**Comment by stdweird** [comment]
with lstat it now works like ln
```
$ PERL5LIB=/usr/lib/perl/ perl -e 'use LC::Check;LC::Check::symlink("/tmp/broken", "missing_pokemon", nocheck => 1);LC::Check::hardlink("/tmp/hardlink_of_broken_symlink", "/tmp/broken")'
created symlink /tmp/broken -> missing_pokemon
updated hard link /tmp/hardlink_of_broken_symlink => /tmp/broken
```
nice work!

**Comment by stdweird** [comment]
The only issue left is that the `VERSION` in the module needs an update, so we can use the version with the fix in `CAF::Path`.

**Comment by stdweird** [comment]
@ned21 @jrha i'm happy with this. i'm in general reluctant to make changes to LC, but this is ok for me to merge.

**Pull request merged**